File: apps/cargo-tracker/packs/tracking/app/domain/tracking/domain/tracking_status.rb
# frozen_string_literal: true

module Tracking
  module Domain
    # 輸送状態（TrackingStatus）。荷役・手動更新で遷移する（IT5 対象値）。
    # 値等価の不変オブジェクト。
    class TrackingStatus
      NOT_RECEIVED = "NOT_RECEIVED"
      RECEIVED = "RECEIVED"
      LOADED = "LOADED"
      ONBOARD_CARRIER = "ONBOARD_CARRIER"
      UNLOADED = "UNLOADED"
      AWAITING_CLAIM = "AWAITING_CLAIM"
      CLAIMED = "CLAIMED"

      VALUES = [
        NOT_RECEIVED, RECEIVED, LOADED, ONBOARD_CARRIER, UNLOADED, AWAITING_CLAIM, CLAIMED
      ].freeze

      attr_reader :value

      def self.initial = new(value: NOT_RECEIVED)

      def initialize(value:)
        raise ArgumentError, "輸送状態が不正です: #{value}" unless VALUES.include?(value)

        @value = value
        freeze
      end

      def not_received? = value == NOT_RECEIVED
      def claimed? = value == CLAIMED

      def ==(other) = other.is_a?(TrackingStatus) && other.value == value
      alias eql? ==
      def hash = value.hash
      def to_s = value
    end
  end
end

File: docs/article/source/scala/apps/cargo-tracker/app/cargotracker/tracking/domain/model/enums/TrackingStatus.scala
package cargotracker.tracking.domain.model.enums

/** 追跡状態（domain-model.md L710-720）。
  *
  *   - 9 段階の追跡フェーズ。`TrackingActivity.currentStatus()` がイベント履歴から導出する
  *   - IT5 では `NotReceived` のみ初期値として利用。`Received` / `Loaded` 等は US15 で `addEvent` 経由で遷移する
  */
enum TrackingStatus:
  case NotReceived, Received, Loaded, OnboardCarrier, Unloaded, AwaitingClaim, Claimed, InException, Unknown
